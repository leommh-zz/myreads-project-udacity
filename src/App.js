import React from 'react'
import { Route } from 'react-router-dom';
import * as BooksAPI from './BooksAPI'
import './App.css'
import AddBooks from './components/AddBooks';
import HomeBooks from './components/HomeBooks';
import BookDetails from './components/BookDetails';


class BooksApp extends React.Component {

  state={
    books: {
      all: [],
      currentlyReading: [],
      wantToRead: [],
      read: []
    }
  }


  //Update State Book
  defineState = (book, shelf) => {
    BooksAPI.update(book, shelf)
    .then(() => this.fetchAll())
  }

  //Fetch pure data API
  fetchAll = () => {
    BooksAPI.getAll()
    .then(res => {
      this.separateBooks(res);
    })
    .catch(error => console.log(error))
  }

  //Separate state pure data API
  separateBooks = (books) => {
    const currentlyReading = books.filter(book => book.shelf==='currentlyReading');
    const wantToRead = books.filter(book => book.shelf==='wantToRead');
    const read = books.filter(book => book.shelf==='read');
    const all = books;

    this.setState(() => ({ books: { currentlyReading, wantToRead, read, all }}))  
  }



  componentDidMount(){
    this.fetchAll();
  }

  render() {
    // console.log('Estado', this.state)
    const {state: { books }, defineState, separateBooks} = this
    
    return (
      <div className="app">
        
        <Route exact path='/search' render={() =>(
          <AddBooks 
            defineState={defineState} 
            booksUser={books} 
          />
        )}/>

        <Route exact path='/' render={() => (
          <HomeBooks 
            defineState={defineState}
            separateBooks={separateBooks}  
            booksUser={books}  
          />
        )}/>

        <Route exact path='/book/:id' component={ (props) =>  <BookDetails {...props} />  }/>
      </div>
    )
  }
}

export default BooksApp
