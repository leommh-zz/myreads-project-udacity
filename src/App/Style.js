const style = {
    main: {
        backgroundColor: '#605872',
        flex: 1,
    },
}

export default style;