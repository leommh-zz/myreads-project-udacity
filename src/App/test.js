import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render } from 'enzyme';
import App from './index';

describe('[Component] App', () => {
   
    const wrapper = render(
        <BrowserRouter>
            <App /> 
        </BrowserRouter> 
    );

    it('Display App Component Mount', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
});