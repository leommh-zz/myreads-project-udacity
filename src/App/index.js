import React from 'react'
import { Route } from 'react-router-dom';
import * as BooksAPI from '../BooksAPI';
import {Grid} from '@material-ui/core';
import {debounce} from 'throttle-debounce';
import Header from '../components/Header';
import BookAdd from '../pages/BookAdd';
import Home from '../pages/Home';
import Style from './Style';

class BooksApp extends React.Component {

  /**
   * @description - Aplication state
   * @var {object} books - All classified books and their divisions
   * @var {string} searchQuery - Input search content   
   * @var {array} booksSearch - All books searched
   * @var {string} title - Personalized Title of each BookCase
   * @var {boolean} loading - Application load state
   */
  state = {
    books: {
      all: [],
      currentlyReading: [],
      wantToRead: [],
      read: []
    },
    searchQuery: '',
    booksSearch: [],
    title: '',
    loading: true,
  }

  /**
  * @description - Update status book (currentlyReading, wantToRead, read)
  * @param {object} book - Book selected data
  * @param {string} shelf - New status to book (currentlyReading, wantToRead, read)
  * @returns {function} - Execute the function fetchAll and performs the search again to 
  * refresh the updated book in the search screen
  */
  defineStatus = (book, shelf) => {
    BooksAPI.update(book, shelf)
    .then(() => this.fetchAll())
    .then(() => this.search(this.state.searchQuery))
  }
  
  /**
  * @description - Brings all classified (currentlyReading, wantToRead, read) books 
  * @returns {function} - Executes the function separateBooks passing the books brought by API
  */
  fetchAll = () => {
    this.setState({ loading: true })
    BooksAPI.getAll()
    .then(res => {
      this.separateBooks(res);
    })
    .catch(error => console.log(error))
  }

  /**
  * @description - Separate status (currentlyReading, wantToRead, read) of books in arrays
  * @param {object} books - Books 
  * @returns {object} - New state books: 
      @param {array} currentlyReading - All books in the status currentlyReading
      @param {array} wantToRead - All books in the status wantToRead
      @param {array} read - All books in the status read
      @param {array} all - All books in the status currentlyReading, wantToRead and read
  */
  separateBooks = (books) => {
    const currentlyReading = books.filter(book => book.shelf==='currentlyReading');
    const wantToRead = books.filter(book => book.shelf==='wantToRead');
    const read = books.filter(book => book.shelf==='read');
    const all = books;

    this.setState(() => ({books: {currentlyReading, wantToRead, read, all}, loading: false}))  
  }

  /**
  * @description - Search the books
  * @param {string} query - Input search value
  * @returns {function} - If the parameter is empty it will execute the function 
  * fetchNone clean the books and the search screen, otherwise it will execute the 
  * fetchSearch function passing the parameter 
  */
  search = (query) => {
    this.setState({searchQuery: query, loading: true}, () => {
      console.log(this.state.searchQuery);
      const fetchSearch = debounce(500, this.fetchSearch);
      const fetchNone = debounce(500, () => this.setState({ booksSearch: [], loading: false }))
      query === '' ? (
        fetchNone()
      ) : (
        fetchSearch(query)
      )
    });
  }

  /**
    * @description - Use API Search books
    * @param {string} searchQuery - Search query value
    * @returns {function} - If the API succeeds in fetching the query then it will call the 
    * booksAddShelf function and pass the purely searched books, otherwise it will display 
    * an error message updating the state
  */
  fetchSearch = (searchQuery) => {
    BooksAPI.search(searchQuery)
    .then(booksPure => this.booksAddShelf(booksPure))
    .catch(() => this.setState({ booksSearch: {}, title: 'Books not found!', loading: false }))  
  }

  /**
    * @description - Filters the pure books and adds the shelf property to books that have already been sorted by the user
    * @param {array} books - Pure books response of API
    * @returns {array} - Returns a new state for the booksSearch array that now contains the 
    * searched books that have not been sorted and also those that have been sorted
  */
  booksAddShelf = (booksPure) => {
    const { books } = this.state;
    const booksSearch = booksPure.filter( bookPure => books.all.map( book =>
      bookPure.id === book.id ? ( 
        bookPure.shelf = book.shelf 
      ) : (
        bookPure
      )
    ));
    this.setState({ booksSearch, title: '', loading: false });
  }

  /**
    * @description - Search for classified books when the application starts
  */
  componentDidMount(){
    this.fetchAll();
  }

  render() {
    const { state: { books, booksSearch, loading, title }, defineStatus, separateBooks, search } = this
    return (
      <Grid container direction="column" alignItems="center" style={Style.main}>
        <Grid item>
          <Header />
        </Grid>
        <Grid item>
          <Route exact path='/search' render={() =>(
            <BookAdd 
              defineStatus={defineStatus} 
              booksUser={books} 
              loading={loading}
              booksSearch={booksSearch}
              search={search}
              title={title}
            />
          )}/>
        
          <Route exact path='/' render={() => (
            <Home 
              defineStatus={defineStatus}
              separateBooks={separateBooks}  
              booksUser={books}  
              loading={loading}
            />
          )}/>
        </Grid>
      </Grid>
    )
  }
}

export default BooksApp;
