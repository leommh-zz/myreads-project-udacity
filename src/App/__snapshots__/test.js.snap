// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] App Display App Component Mount 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "MuiGrid-container-1 MuiGrid-direction-xs-column-4 MuiGrid-align-items-xs-center-9",
      "style": "background-color:#605872;flex:1",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiGrid-item-2",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiGrid-container-1",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "MuiGrid-item-2",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "MuiTypography-root-98 MuiTypography-subtitle1-116",
                      "style": "font-size:50px;color:#fff",
                    },
                    "children": Array [
                      Object {
                        "data": "MyReads",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "text",
                      },
                    ],
                    "name": "h6",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                      "style": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                      "style": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "MuiGrid-item-2",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiGrid-container-1",
                "style": "background-color:#605872;padding:20px",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "MuiGrid-container-1 MuiGrid-item-2 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18",
                    "style": "flex:1",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "MuiGrid-item-2",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {
                            "class": "MuiCircularProgress-root-133 MuiCircularProgress-colorPrimary-136 MuiCircularProgress-indeterminate-135",
                            "role": "progressbar",
                            "style": "width:40px;height:40px;color:#333",
                          },
                          "children": Array [
                            Object {
                              "attribs": Object {
                                "class": "MuiCircularProgress-svg-138",
                                "viewBox": "22 22 44 44",
                              },
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "class": "MuiCircularProgress-circle-139 MuiCircularProgress-circleIndeterminate-141",
                                    "cx": "44",
                                    "cy": "44",
                                    "fill": "none",
                                    "r": "18.5",
                                    "stroke-width": "7",
                                  },
                                  "children": Array [],
                                  "name": "circle",
                                  "namespace": "http://www.w3.org/2000/svg",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "class": undefined,
                                    "cx": undefined,
                                    "cy": undefined,
                                    "fill": undefined,
                                    "r": undefined,
                                    "stroke-width": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "class": undefined,
                                    "cx": undefined,
                                    "cy": undefined,
                                    "fill": undefined,
                                    "r": undefined,
                                    "stroke-width": undefined,
                                  },
                                },
                              ],
                              "name": "svg",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {
                                "class": undefined,
                                "viewBox": undefined,
                              },
                              "x-attribsPrefix": Object {
                                "class": undefined,
                                "viewBox": undefined,
                              },
                            },
                          ],
                          "name": "div",
                          "namespace": "http://www.w3.org/1999/xhtml",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "class": undefined,
                            "role": undefined,
                            "style": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "class": undefined,
                            "role": undefined,
                            "style": undefined,
                          },
                        },
                      ],
                      "name": "div",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                    "style": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                    "style": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "style": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "MuiGrid-item-2",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "MuiGrid-container-1",
              "style": "background-color:#605872;padding:20px",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "MuiGrid-container-1 MuiGrid-item-2 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18",
                  "style": "flex:1",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "MuiGrid-item-2",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {
                          "class": "MuiCircularProgress-root-133 MuiCircularProgress-colorPrimary-136 MuiCircularProgress-indeterminate-135",
                          "role": "progressbar",
                          "style": "width:40px;height:40px;color:#333",
                        },
                        "children": Array [
                          Object {
                            "attribs": Object {
                              "class": "MuiCircularProgress-svg-138",
                              "viewBox": "22 22 44 44",
                            },
                            "children": Array [
                              Object {
                                "attribs": Object {
                                  "class": "MuiCircularProgress-circle-139 MuiCircularProgress-circleIndeterminate-141",
                                  "cx": "44",
                                  "cy": "44",
                                  "fill": "none",
                                  "r": "18.5",
                                  "stroke-width": "7",
                                },
                                "children": Array [],
                                "name": "circle",
                                "namespace": "http://www.w3.org/2000/svg",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {
                                  "class": undefined,
                                  "cx": undefined,
                                  "cy": undefined,
                                  "fill": undefined,
                                  "r": undefined,
                                  "stroke-width": undefined,
                                },
                                "x-attribsPrefix": Object {
                                  "class": undefined,
                                  "cx": undefined,
                                  "cy": undefined,
                                  "fill": undefined,
                                  "r": undefined,
                                  "stroke-width": undefined,
                                },
                              },
                            ],
                            "name": "svg",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {
                              "class": undefined,
                              "viewBox": undefined,
                            },
                            "x-attribsPrefix": Object {
                              "class": undefined,
                              "viewBox": undefined,
                            },
                          },
                        ],
                        "name": "div",
                        "namespace": "http://www.w3.org/1999/xhtml",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "class": undefined,
                          "role": undefined,
                          "style": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "class": undefined,
                          "role": undefined,
                          "style": undefined,
                        },
                      },
                    ],
                    "name": "div",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "style": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "MuiGrid-item-2",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "class": "MuiGrid-container-1",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "MuiGrid-item-2",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "class": "MuiTypography-root-98 MuiTypography-subtitle1-116",
                        "style": "font-size:50px;color:#fff",
                      },
                      "children": Array [
                        Object {
                          "data": "MyReads",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "text",
                        },
                      ],
                      "name": "h6",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                        "style": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                        "style": undefined,
                      },
                    },
                  ],
                  "name": "div",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
